libs/web3/src/lib/use-ethereum-transaction-store.tsx

```diff
@@ -0,0 +1,180 @@
+import produce from 'immer';
+import type { ethers } from 'ethers';
+import type { EthereumError } from './ethereum-error';
+import { isExpectedEthereumError } from './ethereum-error';
+import { isEthereumError } from './ethereum-error';
+
+import type { MultisigControl } from '@vegaprotocol/smart-contracts';
+import type { CollateralBridge } from '@vegaprotocol/smart-contracts';
+import type { Token } from '@vegaprotocol/smart-contracts';
+import type { TokenFaucetable } from '@vegaprotocol/smart-contracts';
+
+import type { EthTxState } from './use-ethereum-transaction';
+import { EthTxStatus } from './use-ethereum-transaction';
+
+type Contract = MultisigControl & CollateralBridge & Token & TokenFaucetable;
+type ContractMethod =
+  | keyof MultisigControl
+  | keyof CollateralBridge
+  | keyof Token
+  | keyof TokenFaucetable;
+
+interface EthStoredTxState extends EthTxState {
+  id: number;
+  contract: Contract;
+  methodName: ContractMethod;
+  args: string[];
+  requiredConfirmations: number;
+  requiresConfirmation: boolean;
+}
+
+export const useEthTransactionStore = create<{
+  transactions: (EthStoredTxState | undefined)[];
+  create: (
+    contract: Contract,
+    methodName: ContractMethod,
+    args: string[],
+    requiredConfirmations: number,
+    requiresConfirmation: boolean
+  ) => number;
+  update: (
+    id: EthStoredTxState['id'],
+    update?: Partial<EthStoredTxState>
+  ) => void;
+  get: (id: EthStoredTxState['id']) => EthStoredTxState | undefined;
+}>((set, get) => ({
+  transactions: [] as EthStoredTxState[],
+  update: (id: EthStoredTxState['id'], update?: Partial<EthStoredTxState>) => {
+    set({
+      transactions: produce(get().transactions, (draft) => {
+        const transaction = draft.find((transaction) => transaction?.id === id);
+        if (transaction) {
+          Object.assign(transaction, update);
+        }
+      }),
+    });
+  },
+  get: (id: number) =>
+    get().transactions.find((transaction) => transaction?.id === id),
+  create: (
+    contract: Contract,
+    methodName: ContractMethod,
+    args: string[] = [],
+    requiredConfirmations: number,
+    requiresConfirmation: boolean
+  ) => {
+    const transactions = get().transactions;
+    const transaction: EthStoredTxState = {
+      id: transactions.length,
+      contract,
+      methodName,
+      args,
+      status: EthTxStatus.Default,
+      error: null,
+      txHash: null,
+      receipt: null,
+      confirmations: 0,
+      dialogOpen: false,
+      requiredConfirmations,
+      requiresConfirmation,
+    };
+    set({ transactions: transactions.concat(transaction) });
+    return transaction.id;
+  },
+}));
+
+export const useEthTransactionManager = () => {
+  const update = useEthTransactionStore((state) => state.update);
+  const transaction = useEthTransactionStore((state) =>
+    state.transactions.find(
+      (transaction) => transaction?.status === EthTxStatus.Default
+    )
+  );
+
+  if (transaction) {
```

Comment:
I mentioned before, but I don't love triggering a tx off the back of a state change. We should be able to trigger it from an event within the app (IE onClick)